pkg/pod/status.go

```diff
@@ -266,15 +279,45 @@ func createMessageFromResults(results []result.RunResult) (string, error) {
 	return string(bytes), nil
 }
 
+// findStepResultsFetchedByTask fetches step results that the Task needs.
+// It accepts a step name and the TaskResults as input and outputs
+// a map with the name of the step result as the key and the name of the task result that is fetching it as value.
+func findStepResultsFetchedByTask(stepName string, specResults []v1.TaskResult) (map[string]string, error) {
+	neededStepResults := map[string]string{}
+	for _, r := range specResults {
+		if r.Value != nil {
+			if r.Value.StringVal != "" {
+				sName, resultName, err := v1.ExtractStepResultName(r.Value.StringVal)
+				if err != nil {
+					return nil, err
+				}
+				if fmt.Sprintf("%s%s", StepPrefix, sName) == stepName {
+					neededStepResults[resultName] = r.Name
+				}
+			}
+		}
+	}
+	return neededStepResults, nil
+}
+
 // filterResults filters the RunResults and TaskResults based on the results declared in the task spec.
 // It returns a slice of any of the input results that are defined in the task spec, converted to TaskRunResults,
 // and a slice of any of the RunResults that don't represent internal values (i.e. those that should not be displayed in the TaskRun status.
-func filterResults(results []result.RunResult, specResults []v1.TaskResult) ([]v1.TaskRunResult, []result.RunResult) {
+func filterResults(results []result.RunResult, specResults []v1.TaskResult, neededStepResults map[string]string) ([]v1.TaskRunResult, []v1.TaskRunResult, []result.RunResult) {
```

Comment:
Is `neededStepResults` the right thing to pass in here?

IIUC this is the set of results that the Task is interested in, which means we will filter StepResults that aren't exported by the Task, even if the Step spec declares them? We want to make sure that the StepState.Results is populated, regardless of whether the Task chooses to export the values or not.

Reply:
Hmm. I kept it similar to TaskRun Results. Currently, there, we don't surface all the results produced by the Task. We only provide the ones that it declares in its definition.

e.g. in the task below, the result `bar` is not present in the TaskRun status. We only surface `foo` since it was declared by the Task.

```
kind Task
spec:
  results:
    - name: foo
  steps:
    - image: bash
       script: |
         echo hi >> $(results.foo.path)
         echo hello >> $(results.bar.path) 
```

Reply:
Fixed. Introduced StepResults as a type in Steps to resolve this. It is no longer needed now.